### ISSUE TITLE
How efficient is `memory_efficient_attention` compared to flash attention?

### ISSUE THREAD

**Issue by vedantroy** [facebookresearch/xformers#666]
# ❓ Questions and Help

TL;DR: Compared to flash attention, how much memory savings does xformers's memory_efficient_attention give? 10% less savings? 20% less savings?

For a vision transformer with:
- 12 layers
- 768 width
- 12 attention heads
- MLP expand/collapse ratio of 4
- Input = 224x224 image split into 16x16 frames

I get the a batch size of 75 images on a V100-16GB with normal attention.
Using `xformers.ops.memory_efficient_attention`, I get a batch size of 110 images.

I don't have an exact comparison, since flash attention doesn't work on V100s, but it seems like flash attention is more memory efficient. Just want to confirm, since if this is not the case, I might have bugs in my code.

**Comment by fmassa** [comment]
Hi,

`memory_efficient_attention` should use the same amount of memory as `flash_attention`, as they both use a similar algorithm.

The savings from using `memory_efficient_attention` will be much more pronounced as you increase the sequence length, e.g., from going from patches of size 16x16 to patches of size 8x8.

I believe I've answered your question, and as such I'm closing this issue. Let us know if you have further questions
